### ISSUE TITLE
Cluster-level hash policy for sticky routing

### ISSUE THREAD

**Issue by agrawroh** [envoyproxy/envoy#23060]
## Description

Currently, the hashing policy is defined on the per-route basis [[Link]](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto.html?highlight=hash_policy#envoy-v3-api-msg-config-route-v3-routeaction-hashpolicy) using `hash_policy`. 

We have a use-case where we want to do sticky routing for all the incoming traffic for the external [ExtAuthZ](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto#extensions-filters-http-ext-authz-v3-extauthz) and [RateLimit](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ratelimit/v3/rate_limit.proto#envoy-v3-api-msg-extensions-filters-http-ratelimit-v3-ratelimit) services but, there is no good way to achieve it. 

We can benefit a lot from a consistent hash Load Balancing like [Ring Hash](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers#ring-hash) etc. by hashing on one of the HTTP headers and leveraging the in-memory cache we have in these upstream services.

Is it possible to support LB hashing policy on the cluster-level (for all the routes)?

**Comment by htuch** [comment]
Yeah, this is support in the `AsyncClient` interface (via `RequestOptions`), but not configurable in a uniform way for things like ext_authz. One option would be to add this to the `GrpcService.EnvoyGrpc` config. This would avoid any major changes such as having to mix routing logic with `ClusterManager`. Does this work?

If not, I think the idea of having some cluster-wide control might have merit, but is a deeper discussion that would require @envoyproxy/api-shepherds and @mattklein123 to weigh in.

**Comment by agrawroh** [comment]
@htuch Thanks for chiming in. There was a [similar request](https://github.com/envoyproxy/envoy/issues/21131) for mirroring traffic. 
Is it possible to identify some of the things that we currently only have on routes which would also make sense to be on the cluster-level and then think more on what would be the best place?

**Comment by htuch** [comment]
Yeah, there are others, e.g. fault injections. One thing I can offer here is a workaround - you can loopback the ext_authz cluster through a listener bound to localhost and have that apply a standard route table before hitting the real backend cluster. This is a total kludge, but if it helps your use case it might be worth considering.

I think we should leave this issue open to gauge wider interest. The API design here would require some careful thought.

**Comment by agrawroh** [comment]
Thanks, @htuch! That's exactly what we are doing right now for mirroring & splitting the traffic :) 

One more question if you know the answer on top of your head - If we have `hash_policy` defined on the routes and the clusters to which the traffic is being mirrored/split have one of `RING_HASH` or `MAGLEV` LB Policy then would it do sticky routing? Or would it ignore the `hash_policy` and the split will be completely random?

**Comment by htuch** [comment]
In both cases Envoy is using an independent per-cluster HTTP async client with its own pseudo-config, so I strongly suspect the answer is it will ignore the original route hash policy.

**Comment by agrawroh** [comment]
@phamann, for latency-sensitive workloads, this might not be the best approach. However, as a workaround, you could open another listener on a new port (say, 9001) with a `/` route and direct your **ext_authz** upstream to this local listener. The actual **ext_authz** upstream would then become the upstream for the `/` route in your new listener. 

This setup would allow you to implement features like hashing, mirroring, etc., on the new route.
